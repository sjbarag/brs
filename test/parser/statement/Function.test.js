const { Lexeme, BrsTypes, Parser } = require("brs");
const { BrsString, Int32 } = BrsTypes;
const BrsError = require("../../../lib/Error");

const { EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("function declarations", () => {
        it("parses minimal empty function declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses non-empty function declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.Print, text: "print", line: 2 },
                { kind: Lexeme.String, text: "Lorem ipsum", line: 2, literal: new BrsString("Lorem ipsum") },
                { kind: Lexeme.Newline, text: "\\n", line: 2 },
                { kind: Lexeme.EndFunction, text: "end function", line: 3 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses functions with implicit-dynamic arguments", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "add2", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },
                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses functions with typed arguments", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "repeat", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.Identifier, text: "str", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "string", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },
                { kind: Lexeme.Identifier, text: "count", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses functions with default argument expressions", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "add", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },

                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "3", line: 1, literal: new Int32(3) },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "4", line: 1, literal: new Int32(4) },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "c", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Plus, text: "+", line: 1 },
                { kind: Lexeme.Integer, text: "5", line: 1, literal: new Int32(5) },
                { kind: Lexeme.RightParen, text: ")", line: 1 },

                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses functions with typed arguments and default expressions", () => {
             let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "add", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },

                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "3", line: 1, literal: new Int32(3) },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Plus, text: "+", line: 1 },
                { kind: Lexeme.Integer, text: "5", line: 1, literal: new Int32(5) },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },

                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses return types", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "void", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });

    describe("sub declarations", () => {
        it("parses minimal sub declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "bar", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndSub, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses non-empty sub declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.Print, text: "print", line: 2 },
                { kind: Lexeme.String, text: "Lorem ipsum", line: 2, literal: new BrsString("Lorem ipsum") },
                { kind: Lexeme.Newline, text: "\\n", line: 2 },
                { kind: Lexeme.EndSub, text: "end sub", line: 3 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses subs with implicit-dynamic arguments", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "add2", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },
                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses subs with typed arguments", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "repeat", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.Identifier, text: "str", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "string", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },
                { kind: Lexeme.Identifier, text: "count", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses subs with default argument expressions", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "add", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },

                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "3", line: 1, literal: new Int32(3) },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "4", line: 1, literal: new Int32(4) },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "c", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Plus, text: "+", line: 1 },
                { kind: Lexeme.Integer, text: "5", line: 1, literal: new Int32(5) },
                { kind: Lexeme.RightParen, text: ")", line: 1 },

                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndSub, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses subs with typed arguments and default expressions", () => {
             let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "add", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },

                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Integer, text: "3", line: 1, literal: new Int32(3) },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.Comma, text: ",", line: 1 },

                { kind: Lexeme.Identifier, text: "b", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Identifier, text: "a", line: 1 },
                { kind: Lexeme.Plus, text: "+", line: 1 },
                { kind: Lexeme.Integer, text: "5", line: 1, literal: new Int32(5) },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },

                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndSub, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("doesn't allow return types", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Identifier, text: "as", line: 1 },
                { kind: Lexeme.Identifier, text: "integer", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndSub, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeTruthy();
        });
    });
});
