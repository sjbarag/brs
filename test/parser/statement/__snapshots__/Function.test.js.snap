// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser function declarations parses functions with default argument expressions 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 3,
            },
          },
          "name": "a",
          "type": 9,
        },
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 4,
            },
          },
          "name": "b",
          "type": 9,
        },
        Object {
          "defaultValue": Binary {
            "left": Variable {
              "name": Object {
                "kind": 21,
                "line": 1,
                "text": "a",
              },
            },
            "right": Literal {
              "value": Int32 {
                "kind": 3,
                "value": 5,
              },
            },
            "token": Object {
              "kind": 8,
              "line": 1,
              "text": "+",
            },
          },
          "name": "c",
          "type": 9,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
  },
]
`;

exports[`parser function declarations parses functions with implicit-dynamic arguments 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": undefined,
          "name": "a",
          "type": 9,
        },
        Object {
          "defaultValue": undefined,
          "name": "b",
          "type": 9,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add2",
    },
  },
]
`;

exports[`parser function declarations parses functions with typed arguments 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": undefined,
          "name": "str",
          "type": 2,
        },
        Object {
          "defaultValue": undefined,
          "name": "count",
          "type": 3,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "repeat",
    },
  },
]
`;

exports[`parser function declarations parses functions with typed arguments and default expressions 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 3,
            },
          },
          "name": "a",
          "type": 3,
        },
        Object {
          "defaultValue": Binary {
            "left": Variable {
              "name": Object {
                "kind": 21,
                "line": 1,
                "text": "a",
              },
            },
            "right": Literal {
              "value": Int32 {
                "kind": 3,
                "value": 5,
              },
            },
            "token": Object {
              "kind": 8,
              "line": 1,
              "text": "+",
            },
          },
          "name": "b",
          "type": 3,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
  },
]
`;

exports[`parser function declarations parses minimal empty function declarations 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
  },
]
`;

exports[`parser function declarations parses non-empty function declarations 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [
          Print {
            "expressions": Array [
              Literal {
                "value": BrsString {
                  "kind": 2,
                  "value": "Lorem ipsum",
                },
              },
            ],
          },
        ],
      },
      "parameters": Array [],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
  },
]
`;

exports[`parser function declarations parses return types 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [],
      "returns": 10,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
  },
]
`;

exports[`parser sub declarations parses minimal sub declarations 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "bar",
    },
  },
]
`;

exports[`parser sub declarations parses non-empty sub declarations 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [
          Print {
            "expressions": Array [
              Literal {
                "value": BrsString {
                  "kind": 2,
                  "value": "Lorem ipsum",
                },
              },
            ],
          },
        ],
      },
      "parameters": Array [],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
  },
]
`;

exports[`parser sub declarations parses subs with default argument expressions 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 3,
            },
          },
          "name": "a",
          "type": 9,
        },
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 4,
            },
          },
          "name": "b",
          "type": 9,
        },
        Object {
          "defaultValue": Binary {
            "left": Variable {
              "name": Object {
                "kind": 21,
                "line": 1,
                "text": "a",
              },
            },
            "right": Literal {
              "value": Int32 {
                "kind": 3,
                "value": 5,
              },
            },
            "token": Object {
              "kind": 8,
              "line": 1,
              "text": "+",
            },
          },
          "name": "c",
          "type": 9,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
  },
]
`;

exports[`parser sub declarations parses subs with implicit-dynamic arguments 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": undefined,
          "name": "a",
          "type": 9,
        },
        Object {
          "defaultValue": undefined,
          "name": "b",
          "type": 9,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add2",
    },
  },
]
`;

exports[`parser sub declarations parses subs with typed arguments 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": undefined,
          "name": "str",
          "type": 2,
        },
        Object {
          "defaultValue": undefined,
          "name": "count",
          "type": 3,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "repeat",
    },
  },
]
`;

exports[`parser sub declarations parses subs with typed arguments and default expressions 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "statements": Array [],
      },
      "parameters": Array [
        Object {
          "defaultValue": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 3,
            },
          },
          "name": "a",
          "type": 3,
        },
        Object {
          "defaultValue": Binary {
            "left": Variable {
              "name": Object {
                "kind": 21,
                "line": 1,
                "text": "a",
              },
            },
            "right": Literal {
              "value": Int32 {
                "kind": 3,
                "value": 5,
              },
            },
            "token": Object {
              "kind": 8,
              "line": 1,
              "text": "+",
            },
          },
          "name": "b",
          "type": 3,
        },
      ],
      "returns": 9,
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
  },
]
`;
